runtime/src/accounts_db.rs

```diff
@@ -2876,59 +2975,102 @@ impl AccountsDB {
                     "Flushing random slot: {}, num_remaining: {}",
                     *rand_slot, num_slots_remaining
                 );
-                self.flush_slot_cache(*rand_slot);
+                self.flush_slot_cache(*rand_slot, None::<&mut fn(&_, &_) -> bool>, false);
             }
         }
-
-        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
     }
 
-    fn flush_slot_cache(&self, slot: Slot) {
+    fn flush_slot_cache(
+        &self,
+        slot: Slot,
+        mut should_flush_f: Option<&mut impl FnMut(&Pubkey, &Account) -> bool>,
+        is_root: bool,
+    ) -> bool {
         info!("flush_slot_cache slot: {}", slot);
-        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+        let slot_cache = self.accounts_cache.slot_cache(slot);
+        let slot_exists_in_cache = slot_cache.is_some();
+        if let Some(slot_cache) = slot_cache {
             let iter_items: Vec<_> = slot_cache.iter().collect();
             let mut total_size = 0;
+            let mut purged_slot_pubkeys: HashSet<(Slot, Pubkey)> = HashSet::new();
+            let mut pubkey_to_slot_set: Vec<(Pubkey, Slot)> = vec![];
+            let mut is_dead_slot = true;
             let (accounts, hashes): (Vec<(&Pubkey, &Account)>, Vec<Hash>) = iter_items
                 .iter()
-                .map(|iter_item| {
+                .filter_map(|iter_item| {
                     let key = iter_item.key();
                     let account = &iter_item.value().account;
-                    let hash = iter_item.value().hash;
-                    total_size += (account.data.len() + STORE_META_OVERHEAD) as u64;
-                    ((key, account), hash)
+                    if should_flush_f
+                        .as_mut()
+                        .map(|should_flush_f| should_flush_f(key, account))
+                        .unwrap_or(true)
```

Comment:
nits: in this case, I'd like to put this loong condition into a humble variable named like `let should_flush = `... :)